tests/conftest.py

```diff
@@ -47,31 +47,66 @@ def pytest_configure(config):
     )
 
 
+def _get_worker_id(config):
+    """
+    Returns the worker id for the current worker if running with pytest-xdist.
+    None if pytest-xdist is not installed or not running in parallel.
+    """
+    # Extract the worker id using internal pytest APIs.
+    if not hasattr(config, "workerinput"):
+        return None
+
+    # workerid has format 'gw0', 'gw1', etc.
+    worker_id = config.workerinput["workerid"]
+    if not worker_id.startswith("gw"):
+        return None
+
+    # skip the 'gw' prefix.
+    return int(worker_id[2:])
+
+
 def _set_default_device(config):
+    """
+    Distributes the tests over multiple GPUs.
+    """
     distribute = int(config.getoption("--gpu-distribute"))
     if distribute < 1:
         return
 
-    if not hasattr(config, "workerinput"):
+    worker_id = _get_worker_id(config)
+    if worker_id is None:
         return
 
-    worker_id = config.workerinput["workerid"]
-    if not worker_id.startswith("gw"):
-        return
-
-    device_id = int(worker_id[2:]) % int(distribute)
+    device_id = worker_id % distribute
 
     import iree.turbine.kernel.wave.utils.general_utils as general_utils
 
     general_utils.DEFAULT_GPU_DEVICE = device_id
 
 
+def _set_cache_dir(config):
+    """
+    Sets the unique cache directory for the current worker to avoid race conditions.
+    """
```

Comment:
What sorts of race conditions? If tests have race conditions, user applications may too. Can the cache be made more multi-thread or multi-process friendly?

Reply:
wave runtime path is writing temp files into the cache dir with potentially colliding names, so you run multiple instances running in parallel you will get into the troubles.